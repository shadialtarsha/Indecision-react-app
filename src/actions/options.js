import { database } from '../firebase/firebase';

export const addOption = option => ({
  type: 'ADD_OPTION',
  option,
});

export const startAddOption = option => (dispatch, getState) => {
  const { auth: { uid } } = getState();
  return database
    .ref(`users/${uid}/options`)
    .push({ text: option })
    .then(ref => {
      dispatch(
        addOption({
          id: ref.key,
          text: option,
        })
      );
    })
    .catch(e => console.log(e));
};

export const removeOption = id => ({
  type: 'REMOVE_OPTION',
  id,
});

export const startRemoveOption = id => (dispatch, getState) => {
  const { auth: { uid } } = getState();
  return database
    .ref(`users/${uid}/options/${id}`)
    .remove()
    .then(() => {
      dispatch(removeOption(id));
    })
    .catch(e => console.log(e));
};

export const removeAll = () => ({
  type: 'REMOVE_ALL',
});

export const startRemoveAll = () => (dispatch, getState) => {
  const { auth: { uid } } = getState();
  return database
    .ref(`users/${uid}/options`)
    .remove()
    .then(() => {
      dispatch(removeAll());
    })
    .catch(e => console.log(e));
};

export const setOptions = options => ({
  type: 'SET_OPTIONS',
  options,
});

export const startSetOptions = () => (dispatch, getState) => {
  const { auth: { uid } } = getState();
  return database
    .ref(`users/${uid}/options`)
    .once('value')
    .then(snapshot => {
      const options = [];
      snapshot.forEach(childSnapshot => {
        options.push({
          id: childSnapshot.key,
          text: childSnapshot.val().text,
        });
      });
      dispatch(setOptions(options));
    })
    .catch(e => console.log(e));
};
