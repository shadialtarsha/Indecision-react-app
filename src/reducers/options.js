import getOptions from '../utils/getOptions';
import setOptions from '../utils/setOptions';

const optionsReducerDefaultState = getOptions('options');

const optionsReducer = (state = optionsReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_OPTION': {
      const newState = [...state, action.option];
      setOptions('options', newState);
      return newState;
    }
    case 'REMOVE_OPTION': {
      const newState = state.filter(option => option !== action.option);
      setOptions('options', newState);
      return newState;
    }
    case 'REMOVE_ALL': {
      setOptions('options', []);
      return [];
    }
    default:
      return state;
  }
};

export default optionsReducer;
