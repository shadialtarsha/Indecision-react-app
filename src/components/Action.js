import React from 'react';
import { connect } from 'react-redux';
import { selectOption } from '../actions/selectedOption';
import pickOption from '../selectors/pickOption';

const Action = props => (
  <div>
    <button
      className="big-button"
      onClick={() => {
        props.selectOption(pickOption(props.options));
      }}
      disabled={props.options.length === 0}
    >
      What Should I do?
    </button>
  </div>
);

const mapDispatchToProps = dispatch => ({
  selectOption: option => dispatch(selectOption(option)),
});

const mapStateToProps = state => ({
  options: state.options,
});

export default connect(mapStateToProps, mapDispatchToProps)(Action);
