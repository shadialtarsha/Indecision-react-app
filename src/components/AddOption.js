import React, { Component } from 'react';
import { connect } from 'react-redux';
import { addOption } from '../actions/options';

export class AddOption extends Component {
  state = {
    option: '',
    error: undefined,
  };

  onInputChange = event => {
    const option = event.target.value;
    this.setState(() => ({ option }));
  };

  handleAddOption = event => {
    event.preventDefault();
    const option = this.state.option.trim();
    let error;
    if (!option) {
      error = 'Enter valid value to add item';
    } else if (this.props.options.indexOf(option) > -1) {
      error = 'This option is already exists';
    } else {
      this.props.onAddOption(option);
    }
    this.setState(() => ({ error }));
    if (!error) {
      this.setState(() => ({ option: '' }));
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p className="add-option-error">{this.state.error}</p>}
        <form className="add-option-form" onSubmit={this.handleAddOption}>
          <input
            className="add-option-form__input"
            value={this.state.option}
            onChange={this.onInputChange}
            type="text"
            name="option"
            placeholder="Your option"
          />
          <button className="button">Add Option</button>
        </form>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  options: state.options,
});

const mapDispatchToProps = dispatch => ({
  onAddOption: option => dispatch(addOption(option)),
});

export default connect(mapStateToProps, mapDispatchToProps)(AddOption);
