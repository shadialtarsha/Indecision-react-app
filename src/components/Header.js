import React from 'react';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = props => (
  <div className="header">
    <div className="container header__container">
      <div>
        <h1 className="header__title">{props.title}</h1>
        {props.subtitle && <h2 className="header__subtitle">{props.subtitle}</h2>}
      </div>
      <button className="button button--link" onClick={props.startLogout}>
        Logout
      </button>
    </div>
  </div>
);
Header.defaultProps = {
  title: 'Indecision',
};

const mapDispatchToProps = dispatch => ({
  startLogout: () => dispatch(startLogout()),
});

export default connect(undefined, mapDispatchToProps)(Header);
