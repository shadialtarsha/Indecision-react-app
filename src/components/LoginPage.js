import React from 'react';
import { connect } from 'react-redux';
import { startLogin } from '../actions/auth';

export const LoginPage = props => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Indecision</h1>
      <p>Put your life in the hands of computer</p>
      <button className="button" onClick={props.startLogin}>
        Login with Google
      </button>
    </div>
  </div>
);
const mapDisptachToProps = dispatch => ({
  startLogin: () => dispatch(startLogin()),
});
export default connect(undefined, mapDisptachToProps)(LoginPage);
