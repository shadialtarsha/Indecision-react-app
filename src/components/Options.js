import React from 'react';
import { connect } from 'react-redux';

import Option from './Option';
import { removeAll } from '../actions/options';

export const Options = props => (
  <div>
    <div className="widget-header">
      <h3 className="widget-header__title">Your Options</h3>
      <button
        onClick={() => {
          props.onRemoveAll();
        }}
        className="button button--link"
      >
        RemoveAll
      </button>
    </div>
    {props.options.length === 0 && <p className="widget__message">Please add an option to get started!</p>}
    <div>
      {props.options.map((option, index) => (
        <Option key={option} count={index + 1} optionText={option} onRemove={props.onRemove} />
      ))}
    </div>
  </div>
);

const mapStateToProps = state => ({
  options: state.options,
});

const mapDispatchToProps = dispatch => ({
  onRemoveAll: () => dispatch(removeAll()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Options);
