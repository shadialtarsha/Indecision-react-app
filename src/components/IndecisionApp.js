import React from 'react';

import Header from './Header';
import Action from './Action';
import AddOption from './AddOption';
import Options from './Options';
import OptionModal from './OptionModal';

class IndecisionApp extends React.Component {
  state = { options: [], selectedOption: undefined };

  componentDidMount() {
    try {
      const options = JSON.parse(localStorage.getItem('options'));
      if (options) {
        this.setState(() => ({ options }));
      }
    } catch (e) {
      console.log(e);
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.options.length !== this.state.options.length) {
      localStorage.setItem('options', JSON.stringify(this.state.options));
    }
  }

  handlePick = () => {
    const randomNumber = Math.floor(Math.random() * this.state.options.length);
    const option = this.state.options[randomNumber];
    this.setState(() => ({ selectedOption: option }));
  };

  handleDeleteSelectedOption = () => this.setState(() => ({ selectedOption: undefined }));

  handleDeleteOptions = () => this.setState(() => ({ options: [] }));

  handleDeleteOption = optionToRemove =>
    this.setState(prevState => ({ options: prevState.options.filter(option => option !== optionToRemove) }));

  handleAddOption = option => {
    if (!option) {
      return 'Enter valid value to add item';
    } else if (this.state.options.indexOf(option) > -1) {
      return 'This option is already exists';
    }
    this.setState(prevState => ({ options: prevState.options.concat(option) }));
  };

  render() {
    const subtitle = 'Put your life in the hands of computer';
    return (
      <div>
        <Header subtitle={subtitle} />
        <div className="container">
          <Action hasOptions={this.state.options.length > 0} onHandlePick={this.handlePick} />
          <div className="widget">
            <Options
              options={this.state.options}
              onRemoveAll={this.handleDeleteOptions}
              onRemove={this.handleDeleteOption}
            />
            <AddOption onAddOption={this.handleAddOption} />
          </div>
        </div>
        <OptionModal
          selectedOption={this.state.selectedOption}
          onDeleteSelectedOption={this.handleDeleteSelectedOption}
        />
      </div>
    );
  }
}

export default IndecisionApp;
