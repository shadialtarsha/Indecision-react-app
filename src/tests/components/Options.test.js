import React from 'react';
import { shallow } from 'enzyme';
import { Options } from '../../components/Options';
import options from '../fixtures/options';

test('should render Options correctly with no options', () => {
  const wrapper = shallow(<Options options={[]} />);
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find('p').text()).toBeTruthy();
});

test('should render Options correctly with options data', () => {
  const wrapper = shallow(<Options options={options} />);
  expect(wrapper).toMatchSnapshot();
});

test('should remove all options on click', () => {
  const onRemoveAllSpy = jest.fn();
  const wrapper = shallow(<Options options={options} onRemoveAll={onRemoveAllSpy} />);
  wrapper.find('button').simulate('click');
  expect(onRemoveAllSpy).toHaveBeenCalledWith();
});
