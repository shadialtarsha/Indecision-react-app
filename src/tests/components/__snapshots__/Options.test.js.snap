// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Options correctly with no options 1`] = `
<div>
  <div
    className="widget-header"
  >
    <h3
      className="widget-header__title"
    >
      Your Options
    </h3>
    <button
      className="button button--link"
      onClick={[Function]}
    >
      RemoveAll
    </button>
  </div>
  <p
    className="widget__message"
  >
    Please add an option to get started!
  </p>
  <div />
</div>
`;

exports[`should render Options correctly with options data 1`] = `
<div>
  <div
    className="widget-header"
  >
    <h3
      className="widget-header__title"
    >
      Your Options
    </h3>
    <button
      className="button button--link"
      onClick={[Function]}
    >
      RemoveAll
    </button>
  </div>
  <div>
    <Connect(Option)
      count={1}
      key="Buy a car"
      optionText="Buy a car"
    />
    <Connect(Option)
      count={2}
      key="Go Shopping"
      optionText="Go Shopping"
    />
    <Connect(Option)
      count={3}
      key="Swimming"
      optionText="Swimming"
    />
  </div>
</div>
`;
